### ISSUE TITLE
[jvm-packages] Spark repartitionForData can potentially shuffle all data and lose ordering required for ranking objectives  

### ISSUE THREAD

**Issue by ngoyal2707** [dmlc/xgboost#3489]
Training for ranking objective requires to have the spark dataframe be grouped by `group` i.e. the query groups be next to each other. This assumption is at following places in the code:

1)  [Making group data](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L316)
2) [Calculating ranking objective loss](https://github.com/dmlc/xgboost/blob/master/src/objective/rank_obj.cc#L84)

Currently xgboost-spark code repartitions the data [here](https://github.com/dmlc/xgboost/blob/master/jvm-packages/xgboost4j-spark/src/main/scala/ml/dmlc/xgboost4j/scala/spark/XGBoost.scala#L255) to have partitions = num_workers. This can potentially shuffle all data and lose the above ordering and above assumptions can break.

Workaround:
Make sure the `#partitions = num_workers` and make xgboost-spark not repartition the data for you. 

@CodingCat , @hcho3 : My current work around is good for me, but this should be improved inside xgboost-spark code. Any suggestions?

**Comment by hcho3** [comment]
I think we could do `sortBy()` with `numPartitions`, so as to sort by the group ID's.

**Comment by weitian** [comment]
the sortBy() on large dataset is very expensive, and it does not keep the order for the items within the same group.

I used a wrapper class to take the group of training data, and partition on that to workaround the problem.
I have plan to make it more generic and submit a PR later.

Instead of XGBLabeledPoint, I use the following data structures:
  case class IndexedFeaturePair(key: Short, value: String)
  case class IndexedSuggestion(label: Double, features: Array[IndexedFeaturePair])
  case class IndexedQuery(suggestion_count: Int, suggestions: Array[IndexedSuggestion])

The method looks like this:
  def trainDistributed(
                        queriesRDD: RDD[IndexedQuery],
                        params: Map[String, Any],
                        round: Int,
                        nWorkers: Int,
                        trainingCount: LongAccumulator,
                        testingCount: LongAccumulator,
                        obj: ObjectiveTrait = null,
                        eval: EvalTrait = null,
                        useExternalMemory: Boolean = false,
                        missing: Float = Float.NaN): (Booster, Map[String, Array[Float]]) = {

**Comment by hcho3** [comment]
@weitian 
> I used a wrapper class to take the group of training data, and partition on that to workaround the problem.
I have plan to make it more generic and submit a PR later.

That would be very nice. Thanks!

**Comment by weitian** [comment]
The repartition on RDD[XGBLabeledPoint] also has another issue that it does gurantee to keep the items within the same group in the same partition after repartitioning.

**Comment by weitian** [comment]
1. The random split code for train and test dataset in object Watches also break the group.
2. This line "val (trainIter1, trainIter2) = trainPoints.duplicate" to generate group data may cause memory issues with large dataset. Scala Iterator does lazy-evaluation, but duplicated iterator may force it to cache data between the gap of the two iterators. see https://www.scala-lang.org/old/node/4943

I am trying to refactor my custom code to fix all the above issues without break the existing code structure.
